storey/steps/sample.py

```diff
@@ -42,31 +49,59 @@ def __init__(
         self._window_size = window_size
         self._emit_period = emit_period
         self._emit_before_termination = emit_before_termination
-
+        self._per_key_count = dict()
         self._count = 0
         self._last_event = None
+        self._extract_key: Callable[[Event], str] = self._create_key_extractor(key)
+
+    @staticmethod
+    def _create_key_extractor(key) -> Callable:
+        if key is None:
+            return lambda event: None
+        elif callable(key):
+            return key
+        elif isinstance(key, str):
+            if key == '$key':
+                return lambda event: event.key
+            else:
+                return lambda event: event.body[key]
+        else:
+            raise ValueError(f'Unsupported key type {type(key)}')
 
     async def _do(self, event):
         if event is _termination_obj:
             if self._last_event is not None:
                 await self._do_downstream(self._last_event)
             return await self._do_downstream(_termination_obj)
         else:
-            self._count += 1
+            key = self._extract_key(event)
+
+            if key is not None:
+                if key not in self._per_key_count:
+                    self._per_key_count[key] = 1
+                else:
+                    self._per_key_count[key] += 1
+                count = self._per_key_count[key]
+            else:
+                self._count += 1
+                count = self._count
 
             if self._emit_before_termination:
                 self._last_event = event
 
-            if self._should_emit():
+            if self._should_emit(count):
                 self._last_event = None
                 await self._do_downstream(event)
 
-            if self._count == self._window_size:
-                self._count = 0
+            if count == self._window_size:
+                if key is not None:
+                    self._per_key_count[key] = 0
+                else:
+                    self._count = 0
```

Comment:
Not really a regression in this PR, but I think it needs to be as follows to avoid really bad behaviour in case of an error downstream.
```
should_emit = self._should_emit(count)

if count == self._window_size:
  if key is not None:
    self._per_key_count[key] = 0
  else:
    self._count = 0

if should_emit:
  self._last_event = None
  await self._do_downstream(event)
```